### PR TITLE
placement: use struct in order to avoid ignoring changing cache

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?
Previously, we use pointer to store the related variable of `FitRegionCache` which may cause the following bug.

```sh
1. FitRegionCache saves pointer of rules, pointer of region,pointer of stores.
2. rules changed
3. FitRegionCache didn't find the variable is changed as the stored variables are also changed.

```

### What is changed and how it works?
use interval struct to store the required variable in order to avoid the aboved problem. Also, use struct will avoid the nil panic during FitRegionCache.

### Check List

Tests

- Unit test

Side effects

As each `FitRegionCache` will use struct to store the variables, each cache is nearly about `96` byte. If we save 1000k FitRegionCache, it will consume about 91MB memory, thus I think the memory consuming is acceptable.

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tidb-ansible`](https://github.com/pingcap/tidb-ansible):
- Need to cherry-pick to the release branch

### Release note

```release-note
None
```

### CODE REVIEW

File: server/schedule/placement/region_rule_cache.go [hunk 1]
```diff
@@ -77,63 +71,162 @@ func (manager *RegionRuleFitCacheManager) CheckAndGetCache(region *core.RegionIn
...
-			return equalRules(a[i], b[j])
+			return a[i].storeEqual(b[j])
 		})
 	})
 }
 
-func storesEqual(a, b []*core.StoreInfo) bool {
-	if len(a) != len(b) {
+func validateRegion(region *core.RegionInfo) bool {
+	if region == nil || region.GetLeader() == nil || len(region.GetDownPeers()) > 0 || region.GetRegionEpoch() == nil {
```
Comment (rleungx) [inline server/schedule/placement/region_rule_cache.go:128]: directly use return

File: server/schedule/placement/region_rule_cache.go [hunk 2]
```diff
@@ -77,63 +71,162 @@ func (manager *RegionRuleFitCacheManager) CheckAndGetCache(region *core.RegionIn
...
+			id:       rule.ID,
+			group:    rule.GroupID,
+			version:  rule.Version,
+			createTS: rule.CreateTimestamp,
 		})
+	}
+	return c
+}
+
+type storeCache struct {
```
Comment (rleungx) [inline server/schedule/placement/region_rule_cache.go:169]: prefer to use pointer receiver

**Comment by disksing** [comment]
I don't understand. How can a `Rule` be changed? I think we always clone the rule then update it.

File: server/schedule/placement/region_rule_cache.go (hunk 2 above)
Comment (Yisaer) [reply server/schedule/placement/region_rule_cache.go:169]: Why use pointer here? I tend to use struct in order to avoid panic error

**Review by disksing — APPROVED** [review]

File: server/schedule/placement/region_rule_cache.go [hunk 3]
```diff
@@ -77,63 +71,163 @@ func (manager *RegionRuleFitCacheManager) CheckAndGetCache(region *core.RegionIn
...
-			return equalRules(a[i], b[j])
+			return a[i].storeEqual(b[j])
 		})
 	})
 }
 
-func storesEqual(a, b []*core.StoreInfo) bool {
-	if len(a) != len(b) {
+func validateRegion(region *core.RegionInfo) bool {
+	return !(region == nil || region.GetLeader() == nil || len(region.GetDownPeers()) > 0 || region.GetRegionEpoch() == nil)
```
Comment (rleungx) [inline server/schedule/placement/region_rule_cache.go:128]:
```suggestion
	return region != nil && region.GetLeader() != nil && len(region.GetDownPeers()) <= 0 && region.GetRegionEpoch() != nil
```

File: server/schedule/placement/region_rule_cache_test.go [hunk 4]
```diff
@@ -141,18 +144,53 @@ func (s *testRuleSuite) TestRegionRuleFitCache(c *C) {
...
+			name:      "invalid input3",
+			region:    mockRegion(3, 0),
+			rules:     nil,
+			unchanged: false,
+		},
 	}
 	for _, testcase := range testcases {
 		c.Assert(cache.IsUnchanged(testcase.region, testcase.rules, mockStores(3)), Equals, testcase.unchanged)
 	}
+	// Invalid Input3
```
Comment (rleungx) [inline server/schedule/placement/region_rule_cache_test.go:177]: Invalid Input3?

**Review by rleungx — APPROVED** [review]
The rest LGTM

**Pull request merged**